File: src/utils/calculos.ts
import { MATERIAIS } from "@/data/materiais";
import { PERFIS } from "@/data/perfis";
import { Resultado } from "@/types/projeto";

export const calcularPesoEValor = (
  tipoPerfil: string,
  medidas: Record<string, number>,
  materialKey: string,
  quantidade: number = 1
): Resultado | null => {
  const perfil = PERFIS[tipoPerfil];
  const material = MATERIAIS[materialKey];
  if (!perfil || !material) return null;

  let peso: number;
  if (perfil.isVolume) {
    peso = perfil.calcularArea(medidas) * material.densidade;
  } else {
    const areaSecao = perfil.calcularArea(medidas);
    const comprimento = medidas.comprimento / 1000;
    peso = areaSecao * comprimento * material.densidade;
  }

  return {
    pesoUnitario: peso,
    pesoTotal: peso * quantidade,
    valorUnitario: peso * material.precoKg,
    valorTotal: peso * quantidade * material.precoKg,
    material: material.nome,
    precoKg: material.precoKg,
  };
};

export const formatarMedidas = (tipoPerfil: string, medidas: Record<string, number>): string => {
  const perfil = PERFIS[tipoPerfil];
  if (!perfil) return '';
  
  return perfil.campos
    .filter(campo => campo !== 'comprimento')
    .map(campo => `${medidas[campo] || 0}`)
    .join('×') + 'mm';
};

export const formatarDescricaoTecnica = (tipoPerfil: string, medidas: Record<string, number>): string => {
  const perfil = PERFIS[tipoPerfil];
  if (!perfil) return '';
  
  const dimensoes = perfil.campos
    .filter(campo => campo !== 'comprimento')
    .map(campo => `${medidas[campo] || 0}`)
    .join('x');
  
  return `${perfil.nome} ${dimensoes}mm`;
};

File: flutter_app/lib/utils/calculos.dart
import '../models/material_model.dart';
import '../models/perfil_model.dart';
import '../models/projeto_model.dart';

class Calculos {
  /// Calcula peso e valor de um perfil metálico.
  ///
  /// Para perfis lineares: peso = área da seção × comprimento × densidade
  /// Para chapas/volume: peso = volume × densidade
  static Resultado calcular({
    required PerfilMetal perfil,
    required Map<String, double> medidas,
    required MaterialMetal material,
    required int quantidade,
  }) {
    final area = perfil.calcularArea(medidas);

    double pesoUnitario;
    if (perfil.isVolume) {
      // Para chapas, a função já retorna o volume em m³
      pesoUnitario = area * material.densidade;
    } else {
      // Para perfis lineares, multiplica área (m²) × comprimento (m)
      final comprimento = (medidas['comprimento'] ?? 0) / 1000;
      pesoUnitario = area * comprimento * material.densidade;
    }

    final pesoTotal = pesoUnitario * quantidade;
    final valorUnitario = pesoUnitario * material.precoKg;
    final valorTotal = pesoTotal * material.precoKg;

    return Resultado(
      pesoUnitario: pesoUnitario,
      pesoTotal: pesoTotal,
      valorUnitario: valorUnitario,
      valorTotal: valorTotal,
      materialNome: material.nome,
      precoKg: material.precoKg,
    );
  }

  /// Formata peso em kg com 3 casas decimais
  static String formatarPeso(double peso) {
    return '${peso.toStringAsFixed(3)} kg';
  }

  /// Formata valor em R$ com 2 casas decimais
  static String formatarValor(double valor) {
    return 'R\$ ${valor.toStringAsFixed(2)}';
  }
}
